### PR TITLE
Fix undefined behavior with `TorchTrainStep`

### PR DESCRIPTION
**Pull request by epwalsh**
This would replace #364, though I think we should port the tests
over from that.

@dirkgr I think this addresses all of the edge cases we talked about.

### CODE REVIEW

File: tango/integrations/torch/train.py [hunk 1]
```diff
@@ -466,12 +487,12 @@ def is_best_checkpoint() -> bool:
...
-        if val_metric is not None and best_val_metric is not None:
-            return (config.minimize_val_metric and val_metric <= best_val_metric) or (
-                not config.minimize_val_metric and val_metric >= best_val_metric
+        if val_metric is not None and best_val_metric_checkpointed is not None:
+            return (config.minimize_val_metric and val_metric <= best_val_metric_checkpointed) or (
+                not config.minimize_val_metric and val_metric >= best_val_metric_checkpointed
             )
         elif best_batch_loss is not None:
-            return best_batch_loss <= batch_loss
+            return batch_loss <= best_batch_loss
```
Comment (dirkgr) [inline tango/integrations/torch/train.py:495]: I'm not sure we should keep the logic that keeps the best checkpoint based on batch loss. Batch loss is so noisy this is essentially random.

Comment (dirkgr) [reply tango/integrations/torch/train.py:495]: Maybe "total loss since the last time we checkpointed". That would be less random. But that feels like a whole new feature.

Comment (dirkgr) [reply tango/integrations/torch/train.py:495]: If we need some always-available fallback metric for model quality, I think recency will be better than "loss on the last batch before we checkpointed".

**Review by dirkgr — CHANGES_REQUESTED** [review]
FWIW, I have hit this case now in a real scenario as well. I didn't debug deeply, but you can get there even without early stopping.

**Comment by dirkgr** [comment]
I ported the test over.

File: tango/integrations/torch/train.py (hunk 1 above)
Comment (epwalsh) [reply tango/integrations/torch/train.py:495]: Hmm yea using recency is probably okay. bfd336c

Comment (dirkgr) [reply tango/integrations/torch/train.py:495]: I don't think you pushed that.

Comment (epwalsh) [reply tango/integrations/torch/train.py]: Oops, yea push failed. Just pushed now

File: tango/integrations/torch/train.py [hunk 2]
```diff
@@ -484,38 +481,35 @@ def _train(
 
     def is_best_checkpoint() -> bool:
         """
-        A closure that we'll call when saving checkpoints to check if we should hardlink
+        A closure that we'll call when saving checkpoints to check if we should link
```
Comment (dirkgr) [inline tango/integrations/torch/train.py:484]: To be sure, this has to be a hardlink, because the checkpoint that's saved under a name that derives from the step count might go away.

I guess we could do it the other way, where the current best checkpoint becomes a real file, with a symlink pointing to it. Then it doesn't have to be a hardlink anymore. But either way it doesn't work on Windows, so why bother?

**Review by dirkgr — APPROVED** [review]

**Comment by dirkgr** [comment]
Argh, this is a real test failure now.
